### DISCUSSION TITLE
Support LNURL (pay and withdraw)

### DISCUSSION THREAD

**Discussion by talvasconcelos** [btcpayserver/btcpayserver#1692] [General]
**Is your feature request related to a problem? Please describe.**
Sending or receiving Lightning payments can often require a bit of set-up and a number of steps. Lnurl aims to simplify many common actions so that they instead require just a click or a QR scan.

**Describe the solution you'd like**
Implement support for [LNURL](https://github.com/btcontract/lnurl-rfc) 

**Provide examples**
LNURL is a bech32-encoded HTTPS/Onion query string which is supposed to help payer interact with payee and thus simplify a number of standard scenarios such as:

- Requesting incoming channels (LNURL-channel)
- Logging in (LNURL-auth)
- Withdrawing funds (LNURL-withdraw)
- Paying for a service (LNURL-pay)

**Additional context**
This will allow for example the creation of LN invoices, either for receive or send that can be printed and scaned! A street musician prints a LNURL-pay QR code, sticks it on his guitar case and people can tip him a few satoshis!

The creation of a faucet, to give some satoshis to friends, bring a printed QR to a friends party and have them take some sats.

**Comment by NicolasDorier** [comment]
I still don't understand where we should put that in btcpay

**Reply by iBobik** [reply]
About withdraw, from the same reason why there are pull payments - send payouts from hot wallets, just in standardised way, so we does not ask employees to copy&paste LN invoice to a web UI, but just send them LNURL-withdraw link/qr, so they can just use their LN wallet to get the money.

**Comment by talvasconcelos** [comment]
> I still don't understand where we should put that in btcpay

I'm not sure, maybe as a service or as an app type?! Or like the Payment Request...
It can be an opt-in feature, for example, like the Transmuter

**Comment by pavlenex** [comment]
I don't understand why this is needed. What this does that BTCPay already doesn't do?

**Comment by talvasconcelos** [comment]
> I don't understand why this is needed. What this does that BTCPay already doesn't do?

The hability to create a static/reusable/printable Lightning Network invoice, for example!

**Comment by chill117** [comment]
> I still don't understand where we should put that in btcpay

The [PoS app](https://docs.btcpayserver.org/Apps/#point-of-sale-app) in btcpayserver could be a great place to add [lnurl-pay](https://github.com/btcontract/lnurl-rfc/blob/master/lnurl-pay.md) support. Here's a real world example [use case](https://twitter.com/_TaxMeIfYouCan_/status/1274458525368975361):
> We used @BtcpayServer PoS app to accept the payments and we managed to make almost all transactions in LN! UX was pretty simple - customer just install the app, put coins or notes to #Bleskomat, scans the QR code with a link, receives money immediately and can pay! @PhoenixWallet

lnurl + offline signing is a powerful combination for mobile applications used by vendors at festivals or conferences.

**Comment by hsjoberg** [comment]
lnurl-pay lets you provide metadata for a transaction.  
Within the lnurl-pay payload, you could provide a URL for the Invoice that the user's wallet conveniently could store alongside the payment.

It's also possible to embed an image which can yield a better experience and clarity for the user.

**Comment by chill117** [comment]
The new Refund feature in btcpayserver looks like a good place to add lnurl-withdraw - high level flow as follows:
* Merchant authorizes a refund on a previously paid invoice
* Customer receives a notification (likely email?) from btcpayserver instance with instructions (e.g "click this link to get your refund")
* Customer clicks link and opens a browser window
* btcpayserver instance displays a page similar to the payment page. But instead of a payment address or invoice, an lnurl-withdraw QR code is displayed
* The customer scans the QR code with an lnurl-withdraw supporting mobile app (there are many)
* The lnurl-withdraw process is completed and the customer has been refunded their BTC via LN

**Reply by iBobik** [reply]
I would prefer to send LNURL link+qr in the email itself, so user will not neet do open webpage, it will open wallet directly.

**Comment by esky33kubernetes** [comment]
Support for LNURL (pay and withdraw) in BTCPay
any ideas on this being a btcpay plug in ??
LNURL sounds very interesting !! & I'm still trying to understand the full benefits.
@rockstardev @dennisreimann 

https://github.com/fiatjaf/lnurl-rfc

https://github.com/fiatjaf/awesome-lnurl

I noticed with https://github.com/apotdevin/thunderhub fragment on btcpay we can use LNURL 👍🏽 
LNURL integration: ln-pay and ln-withdraw are available. Ln-auth soon.

https://lnurl-toolbox.degreesofzero.com
